Changes needed for compatibility with UI project

stefano-dallona opened the pull request:
Changes needed for compatibility with UI project

Commit:
Changes needed for compatibility with UI project

Commit:
Changes needed for compatibility with UI project

LucaVignati commented on plctestbench/node.py:
```diff
@@ -66,9 +66,13 @@ def _save_to_database(self):
     def get_id(self) -> str:
-        return str(hash(self.settings))
+        hash_properties = Settings.__copy__(self.settings)
+        hash_properties.add("parent", self.parent.get_id() if self.parent!=None else "")
```
Perché aggiungi l'hash del parent qui? Non dovrebbe essere più necessario in seguito alle modifiche apportate con questo commit: 67bc636a187a8f374cbb16593eb360c01c021e1a

stefano-dallona replied:
La modifica era precedente alle tue fix, ma per evitare problemi su una cosa così delicata non l'avevo ancora rimossa. Se è ridondante alla luce delle tue fix ignorala pure.

LucaVignati commented on plctestbench/worker.py:
```diff
@@ -1,13 +1,31 @@
\ No newline at end of file
+        super().__init__(settings)
+
+    def run(self) -> None:
```
Serve avere una dummy progress bar per l'OriginalAudio?

stefano-dallona replied:
Esatto

LucaVignati commented on plctestbench/node.py:
```diff
@@ -117,6 +127,10 @@ def get_data(self) -> np.ndarray:
     def get_original_track_node(self) -> OriginalTrackNode:
         return self.root
     
+    def load(self) -> None:
```
Perché tutti questi load?

stefano-dallona replied:
Perché io utilizzo la gerarchia dei nodi del plctestbench per caricare i file audio, in modo da essere sicuro che il risultato sia per definizione lo stesso.

LucaVignati replied:
Di questo è meglio parlarne a voce e magari guardare il codice insieme

LucaVignati replied:
I load verranno rimossi. Al loro posto verrà sviluppata una funzione che accetta un run_id ed una lista di node_id per ritornare un plc_testbench con l'intera struttura dei nodi del run. Quelli specificati nella lista contengono il file a loro associato.

LucaVignati commented on plctestbench/node.py:
```diff
@@ -66,9 +66,13 @@ def _save_to_database(self):
     
     def run(self):
+        if self.worker != None:
+            self.worker.set_uuid(self.get_id())
```
Perché serve settare l'id nel worker?

stefano-dallona replied:
Perché il worker è il componente che notifica l'avanzamento alla gui e per correlare le informazioni al relativo nodo uso il node id

LucaVignati replied:
Nella classe Worker hai accesso alla classe Settings del nodo. Se non sbaglio puoi ricavare il node id calcolando l'hash dei settings.

LucaVignati reviewed (commented):
Complessivamente siamo molto più vicini dell'altra volta, ottimo lavoro

LucaVignati committed:
Removed redundant information from the hash of the settings

The hash of the parent is now stored as a parameter in the settings instance and used to compute the hash, so it is no longer needed to insert it into the hash in the get_id function

LucaVignati committed:
Fix

Commit:
Added optional connection string management to support mongodb atlas

Commit:
Temporarily disabled pybind due to missing ecc_external folder

Commit:
Merge branch 'ui' of github.com:LucaVignati/plc-testbench into ui

LucaVignati commented on plctestbench/settings.py:
```diff
@@ -183,7 +183,7 @@ def __init__(self, max_frequency: float = 4800,
 class DeepLearningPLCSettings(Settings):
 
-    def __init__(self, model_path: str,
+    def __init__(self, model_path: str = "",
```
Il percorso del modello è obbligatorio per questa classe. Non consideriamo ancora questa classe, quando sarà il momento di testarla renderò disponibile un modello da caricare.

LucaVignati committed:
Fixed a bug in the LowCostPLC
- fixed a typo that caused the LowCostPLC to misbehave

LucaVignati commented on plctestbench/data_manager.py:
```diff
@@ -131,7 +139,10 @@ def _save_run_to_database(self):
         self.run['creator'] = self.user['email']
         self.run['created_on'] = datetime.datetime.now()
         self.run['status'] = 'CREATED'
+        self.run['selected_input_files'] = list(map(lambda node: os.path.basename(node.file.path), self.root_nodes))
```
This information is already present in the database. The "workers" array in the runs collection stores 4 lists. The first one contains the settings for the OriginalAudio workers, there you can find the filename of the selected input files

LucaVignati replied:
![screenshot](https://github.com/LucaVignati/plc-testbench/assets/17725043/87801f92-7c04-4357-805e-62c6bf5fd2d7)

LucaVignati replied:
<img width="323" alt="Screenshot 2023-08-11 at 20 29 43" src="https://github.com/LucaVignati/plc-testbench/assets/17725043/62c521d0-aded-498a-bc99-e6e8a97163b9">

LucaVignati committed:
Minor fixes towards the resolution of the PR

LucaVignati commented on plctestbench/file_wrapper.py:
```diff
@@ -157,7 +157,9 @@ def load(self) -> None:
-    pass
+    @staticmethod
+    def to_json():
+        return lambda o : o.__dict__ if hasattr(o, '__dict__') else o
```
Rimuovere questa funzione, è una lambda qualsiasi non legata ad alcun attributo della classe OutputAnalysis

LucaVignati replied:
Scrivere commento che spiega il motivo per cui serve questa funzione.

stefano-dallona replied:
Ho risolto. Puoi rimuovere questa modifica

LucaVignati committed:
Removed the uuid attribute from the Worker class,
substituted by the get_node_id() function.
- now the Node doesn't need to inject the node_id into the Worker during
the execution of the run function

LucaVignati commented on plctestbench/data_manager.py:
```diff
@@ -37,6 +43,7 @@ def __init__(self, testbench_settings: dict, user: dict) -> None:
             ReconstructedTrackNode,
             OutputAnalysisNode
         ]
+        self.progress_monitor = testbench_settings["progress_monitor"] if 'progress_monitor' in testbench_settings.keys() else lambda caller: std_tqdm
```
Invece di passare il progress monitor nei testbench settings propagandolo poi in tutte le classi, dovremmo usare l'import condizionale di Python. Dovremmo probabilmente fare l'import condizionale in un solo file (per esempio utils) e poi importare negli altri file da lì. L'import condizionale si può fare in tanti modi, anche per dire usando una variabile d'ambiente

stefano-dallona replied:
Ottimo. Implementa pure la modifica e poi io mi adeguo. Tanto l'esigenza mi sembra chiara.

LucaVignati replied:
Ho visto che tu avevi una catena di lambda complicata in cui prima veniva chiamato il progress_monitor(self) che ritornava una funzione tipo tqdm la quale veniva chiamata a sua volta con (range(bla bla bla)). Non ho capito perché funzionasse così per cui per il momento ho tolto il primo step. Ora viene chiamato direttamente progress_monitor(range(bla bla bla)). Fammi sapere se va bene così o se hai esigenze diverse, nel caso ci sentiamo magari brevemente così mi spieghi e posso aggiustare il codice qui

LucaVignati committed:
Added pre-trained Deep Learning model
- fixed default DeepLearningPLCSettings to reflect the parameters
  needed for the included pre-trained model to work
- imported tensorflow in the plc_testbench file

LucaVignati committed:
Removed the load() functions from the Node
subclasses
- added default arguments to PLCTestbench, DataManager and
  DatabaseManager classes.
- added checks on arguments

LucaVignati committed:
Added overwritable progress_monitor alias to tqdm
- the management of the tqdm library (or any substitute) is now done at
  import level, the components of the testbench are not actively
  participating anymore.
- added conditional import of tqdm based on the execution environment
  (terminal or jupyter notebook)
- removed the explicit management of the progress progress monitoring
  from all the interested classes
- removed tqdm imports
- removed to_json static method from the OutputAnalysis class

LucaVignati committed:
Removed 'selected_input_files' from the run
dictionary as it is redundant.
The same information is already present in
run['workers'][0][:]['settings']

LucaVignati committed:
Merge branch 'public' into ui

LucaVignati committed:
Changed the interface of progress_monitor to
match the requirements of Stefano Dallona's UI
- progress_monitor is now a lambda that accepts a parameter but it
  ignores it and return tqdm regardless. progress_monitor can be
  overwritten with any lambda function.
- reverted to progress_monitor(self)(range(...)) for compatibility
  reasons with the ui
- replaced some MSEData leftovers with SimpleCalculatorData
- fixed some imports

LucaVignati commented:
Ci siamo quasi! In questo momento il codice di questo branch è identico a quello che ci sarà su public una volta fatto il merge. Per questo motivo manca solo che tu lo testi nel tuo contesto per vedere se ci è sfuggito qualcosa, dopodiché possiamo chiudere questa pull request.

LucaVignati reviewed (commented):
L'utilizzo da Jupyter funziona senza problemi

LucaVignati reviewed (approved):
L'utilizzo da Jupyter funziona senza problemi

LucaVignati committed:
Reverted back to propagating the progress_monitor
from the inputs to each worker.
- the run() function was moved from the PLCTestbench to the DataManager
  class. This is because the progress_monitor is contained in the
  testbench_settings dictionary that is validated in the DataManager
  class.
- The DataManager class now imports the progress_monitor variable from
  the utils.py file.
- The DataManager class now also saves the progress_monitor in each
  settings class before passing them to the nodes.
- the Worker class now retrieves the progress_monitor from the settings
  class. While doing so, it already calls progress_monitor(self).
- each worker class now doesn't need to import the progress_monitor
  from the utils.py file because it is already stored in each instance.
- each worker class now can use the standard TQDM interface because the
  Worker constructor has already called the progress_monitor(self).
- Removed old console messages in the PEAQCalculator
- Reduced the articifial waiting time in PEAQCalculator and
  OriginalAudio from 1s to 0.1s
- The Settings class now supports setting and getting the
  progress_monitor which is persisted across calls to inherit_from() and
  __copy__()

LucaVignati commented:
Manterrei comunque le conversazioni qui, nella pull request su cui stiamo lavorando. Anche da qui puoi commentare linee di codice specifiche se serve. Altrimenti diventa difficile ripescare i commenti che hai fatto su commit specifici.

Commit:
- Added explicit encoding in setup to solve an exception due to an illegal character
- Fixed missing attribute problem in SimpleCalculatorData (_mse)
- Added progress monitor call in node branch to ensure proper progress notification when the elaboration is skipped in favour of persisted data
- Fixed return parameter of SpectralEnergyCalculator to allow proper classification of the metric as a linear one
- Fixed exception in SpectralEnergyCalculator instantiation caused by missing settings (N, hop, etc.)

LucaVignati committed:
Added new dummy_progress_bar() function in the
utils module for readability and consistency.
- dummy_progress_bar is now used in Node.run(), PEAQCalculator.run()
  and OriginalAudio.run()
